### ISSUE TITLE
Support multi site cluster of indexers

### ISSUE THREAD

**Issue by rfaircloth-splunk** [splunk/splunk-operator#44]
In a cluster with multiple availability zones (1.17+) provide the ability to define a indexer cluster where indexers are added to a Splunk Site and meta for affinity to the zone is added to each indexer pod so that all indexer pods of the same site are in the same zone.

**Comment by mikedickey** [comment]
Similar to #51, I believe this is already supported in 0.1.0.

**Comment by rfaircloth-splunk** [comment]
I could not see how but that would be awesome, it would require some amount of logic to assign indexers to specific sites /affinity groups and inject the number of sites into master

**Comment by mikedickey** [comment]
oic, the annotations should work but we're not setting up any sites inside splunk config.

**Comment by mikedickey** [comment]
Note that #41 would be a dependency of this.

**Comment by romain-bellanger** [comment]
Hello,

This actually seems to be already feasible using the downward API from kubernetes.

## Service account configuration

The default service account of the namespace running the cluster must be granted node-reader ClusterRole and pod-reader Role

**Note:** pod-reader is needed as IndexerCluster doesn't provide a way to define environment variable on pods, which is apparently the [only way to expose spec.nodeName](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api).

<details>
  <summary>Click to expand!</summary>

  ```yaml
  ---
  apiVersion: rbac.authorization.k8s.io/v1
  kind: Role
  metadata:
    name: pod-reader
  rules:
  - apiGroups: [""] # "" indicates the core API group
    resources: ["pods"]
    verbs: ["get", "watch", "list"]
  ---
  apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: splunk-node-reader-view-pods
    namespace: splunk
  subjects:
  - kind: ServiceAccount
    name: default
    #apiGroup: rbac.authorization.k8s.io
  roleRef:
    kind: Role
    name: pod-reader
    #apiGroup: rbac.authorization.k8s.io
  ---
  # Requires cluster admin priviledges
  apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: splunk-node-reader-view-nodes
    namespace: splunk
  subjects:
  - kind: ServiceAccount
    name: default
  roleRef:
    kind: ClusterRole
    name: node-reader
  ```
</details>

## Installing the openshift python module in the docker image

This is the client library used by the [ansible k8s lookup](https://docs.ansible.com/ansible/latest/plugins/lookup/k8s.html#requirements) (lookup not specific to openshift). It could be meaningful to have this integrated into the official Splunk docker image.

<details>
  <summary>Click to expand!</summary>

  ```
  FROM splunk/splunk:8.0.4
  USER root
  RUN pip2 install openshift
  USER ansible
  ```
</details>

## Downward API configuration

As explained in previous sections, pod properties are exposed through volumes, as spec.nodeName can only be exposed through environment variable, which is not yet supported by the IndexerCluster CRD.

In the definition of the IndexerCluster:

<details>
  <summary>Click to expand!</summary>

  ```yaml
  ---
apiVersion: enterprise.splunk.com/v1alpha2
kind: IndexerCluster
metadata:
  name: example
  finalizers:
  - enterprise.splunk.com/delete-pvc
spec:
  # The custom image built at previous step
  image: "splunk/splunk:8.0.4_openshift_client"
  volumes:
    - name: podinfo
      downwardAPI:
        # type has to be defined because of bug in OCP 3.11: https://github.com/kubernetes/kubernetes/issues/68466
        # Because of the same bug IndexerCluster deletion also ends up in error when downwardAPI is used and finalizer
        # enterprise.splunk.com/delete-pvc is defined. Removing finalizer allows deleting the idxc.
        type: array
        items:
          - path: pod
            fieldRef:
              fieldPath: metadata.name
          - path: namespace
            fieldRef:
              fieldPath: metadata.namespace
  ```
</details>

## Splunk ansible configuration

Still in the IndexerCluster spec definition for the automatically generated ConfigMap (could also be defined separately):

<details>
  <summary>Click to expand!</summary>

  ```yaml
  ---
spec:
  defaults: |-
    zone_site_map:
      zone-1a: site1
      zone-1b: site2
      zone-1c: site3
    k8s_token: "{{ lookup('file', '/var/run/secrets/kubernetes.io/serviceaccount/token') }}"
    k8s_namespace: "{{ lookup('file', '/mnt/podinfo/namespace') }}"
    k8s_pod_name: "{{ lookup('file', '/mnt/podinfo/pod') }}"
    k8s_node_name: "{{ lookup('k8s', api_key=k8s_token, kind='Pod', namespace=k8s_namespace, resource_name=k8s_pod_name).spec.nodeName }}"
    # In latest version this label is being replaced with topology.kubernetes.io/zone
    k8s_zone: "{{ lookup('k8s', api_key=k8s_token, kind='Node', resource_name=k8s_node_name).metadata.labels['failure-domain.beta.kubernetes.io/zone'] }}"
    splunk:
      # As this is shared between master and indexers:
      # The cluster master pod can't reach it's own API through the ClusterIP Service before the pod is ready
      # Ansible connects to the API at setup time, before the pod is ready, so using localhost
      multisite_master: "{{ (lookup('env', 'SPLUNK_ROLE') == 'splunk_cluster_master') | ternary('localhost', 'splunk-example-cluster-master-service') }}"
      site: "{{ zone_site_map[k8s_zone] }}"
      all_sites: site1,site2,site3
      multisite_replication_factor_origin: 1
      multisite_replication_factor_total: 2
      multisite_search_factor_origin: 1
      multisite_search_factor_total: 2
      idxc:
        search_factor: 2
        replication_factor: 2
  ```
</details>

## Proposed improvements

**Already mentioned in the previous sections**
* Install the openshift python module in the official Splunk docker image, to enable the usage of the k8s ansible lookup
* Enable defining pod environment variables in the IndexerCluster CRD to expose the spec.nodeName field directly. This would also allow to pass additional parameters to ansible through ANSIBLE_EXTRA_FLAGS, especially -vvv to enable debug logs for troubleshooting.

**Muti-part IndexerCluster**

I've done a small development using the indexerClusterRef section which is already defined in all Splunk CRDs, but still unused in the IndexerCluster, in order to plug multiple indexer clusters together.

When indexerClusterRef is defined:
* no cluster master StatefulSet is created
* cluster-master client created to check for the status targets the cluster-master of the referenced IndexerCluster
* the child IndexerCluster reuses the Secret of the referenced IndexerCluster (mounted to indexer pods)
* the clusterMasterPhase of the child IndexerCluster objects is retrieved from the referenced IndexerCluster
* note: as SearchHeadCluster cluster or Standalone only need to reference the master IndexerCluster to search or and forward their internal logs to all the indexers (using indexer discovery from master)

From our perspective, this brings several operability improvements:
* Defining an IndexerCluster controlling only the cluster-master pod (replicas: 0), with dedicated specs:
  * smaller disk sizes on the master which reduces the cost in public cloud
  * different storage class to use remote/shared storage on cluster-master (allowing it to move across nodes) when local storage is required on indexers (serious impact if master on same node with local storage as indexer pods)
  * image version of the master controlled independently from the indexer, which allows to respect the upgrade cycle: master, search heads, indexers
* Defining several groups of indexers connected to the same cluster master:
  * control the upgrades by group of indexers, which could be mapped to sites/zones, which contains the risk of impact of such operations
  * progressively replace indexers by creating a new group with different specs, for instance different storage class or disk sizes (as volume templates apparently can't be changed in StatefulSet)

We would love to contribute, but before submitting a PR (especially need to finalize the unit tests and documentation), we would like to ensure that this is going in the right direction.

**Comment by mikedickey** [comment]
Thanks, @romain-bellanger ! I like your thoughts regarding "Multi-part Indexer Clusters" and re-leveraging indexerClusterRef, as well as allowing environment variables to be passed-through from splunk operator's custom resource specs into the pod specs of the containers (I worry about it making it too easy to break the ansible stuff, but the utility of this has popped up many times before and is probably worth the trade-off).

I would love to hear if you have more thoughts on how configuration for multisite would occur in the specs for custom resources. Perhaps it's as simple as adding a "site" string parameter? Not sure.. I think ideally operator should do all the heavy lifting so that it's as easy to set this up as possible.

I don't follow why using the k8s downward api would be helpful or necessary. You can already use the "affinity" spec to create clusters of indexers that are scheduled into a specific zone.

I'm also not all that up-to-speed on the current state of multisite support inside our docker-splunk container, or more specifically in splunk-ansible. I know there are some parameters for multisite as[ documented here](https://github.com/splunk/splunk-ansible/blob/develop/docs/advanced/default.yml.spec.md), but not sure how much of this we can reuse or would need to change to make the desired [server.conf changes](https://docs.splunk.com/Documentation/Splunk/8.0.4/Indexer/Multisiteconffile). Perhaps @lephino @bb03 @arctan5x or @nwang92 would have more insight on what that would look like?

**Comment by romain-bellanger** [comment]
Hi @mikedickey 

Many thanks for your answer :-)

The purpose of the first part of my previous comment was to show how a multi-site cluster can already be created with the current version of the splunk-operator. That it is already possible with a single IndexerCluster object, with a single/shared spec definition. In that case the downward API is required.

Then with multiple parts, I agree that there are more options. The main problem which I faced when using affinity to map an IndexerCluster part to a zone, was that the PersistentVolume was then generally not created into the same zone, and could not be bound by the pods. I had to ask our cluster admins to create some zone specific StorageClass, which they were a bit reluctant to do. They actually recommended to use the downward API instead. We are working on openshift clusters located in a private Openstack IaaS with local SSD storage, where they agreed to create the zone specific storage classes, and in Azure where our admins answered that the current kubernetes.io/azure-disk provisioner doesn't support allowedTopologies and would have to be migrated to the [CSI driver](https://github.com/kubernetes-sigs/azuredisk-csi-driver).

The multi-part cluster can be used in two ways, it could actually be fine to allow the user to choose:
1) Create a part per site, possibly having to align the affinity and StorageClass settings. Then each site can be operated independently, and the site setting can be hardcoded in the IndexerCluster together with the affinity.
2) Have all the indexers within of a single IndexerCluster object, but allow the indexers to retrieve their zone through the downward API and to pick their site dynamically. This is already supported by the current operator version. As far as I understand, by default the pods of a StatefulSet are placed in round robin across zones. The multi-part cluster still have some value with this solution, to split the cluster-master and indexer specs (benefits: storage size and class, upgrade cycle), but also to have the possibility to connect a new IndexerCluster object with different specs (e.g. new image for upgrades, different storage size) to ease maintenance operations (e.g. progressively replace the previous IndexerCluster object)

On our cluster located on the Openstack IaaS with local storage, we used solution 1. Here are the specs used for the parts:

**The master contains almost all the settings for multi-site setup, and controls the definition of apps:**

Need to use the jinja trick for the multisite_master in this case, as cluster master and indexers have dedicated specs.

<details>
  <summary>Click to expand!</summary>

  ```yaml
  spec:
    replicas: 0
    image: "splunk/splunk:8.0.4"
    volumes:
      - name: apps
        configMap:
          name: example-idxcluster-apps
    defaults: |-
      splunk:
        # The site is not really meaningful for the master in this case
        site: site1
        http_enableSSL: true
        multisite_master: localhost
        all_sites: site1,site2,site3
        multisite_replication_factor_origin: 1
        multisite_replication_factor_total: 2
        multisite_search_factor_origin: 1
        multisite_search_factor_total: 2
        idxc:
          search_factor: 2
          replication_factor: 2
        # Apps defined here are deployed to the indexers of all the other parts
        apps_location:
          - "/mnt/apps/app_test_input1.tgz"
    # The master runs on general purpose nodes with shared storage
    storageClassName: standard
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: node-role.kubernetes.io/compute
              operator: In
              values:
              - "true"
  ```
</details>

**Then each IndexerCluster site only defines the master to which it connects and its site / affinity / storage class:**

<details>
  <summary>Click to expand!</summary>

  ```yaml
  spec:
    replicas: 2
    image: "splunk/splunk:8.0.4"
    defaults: |-
      splunk:
        multisite_master: splunk-example-cluster-master-service
        site: site1
    # The indexers runs on specific nodes with local SSD storage
    storageClassName: local-1a
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: node-role.kubernetes.io/local
              operator: In
              values:
              - "true"
            - key: failure-domain.beta.kubernetes.io/zone
              operator: In
              values:
              - zone-1a
  ```
</details>

I didn't really see any

**Comment by mikedickey** [comment]
This is great work and seems like you've cracked most of it. Much further along that I anticipated!

I think it's fine to just update ValidateIndexerClusterSpec to allow replicas of zero for IndexerCluster since it would become a valid configuration.

Curious about azure having that mismatch between storage volumes and pod scheduling. This seems to be a common / long-standing problem for storage engines, which most have since solved. I hope azure team is working on fixing this bug.

**Comment by romain-bellanger** [comment]
Hi Mike,

That was a mistake about the azure-disk provisioner in my previous comment. This problem of volume not created in the same zone as the pod was experienced in an Openstack environment (VIO, not sure about the version), using Cinder provisioner for local persistent volumes, with an openshift 3.11 / kubernetes 1.11 cluster. Our k8s admins tried to use "volumeBindingMode: WaitForFirstConsumer" on the storage class but this didn't solve the problem, so we had to use a dedicated StorageClass per zone.

The discussion with our k8s admins about Azure took place during the definition of our pre-requisites, as I expected to face the same problem. We only got access recently and I didn't actually test this. I've just tried and asked our k8s admins to double check in which zones the pvs were created. The pv are created in the same zone as the pod, and match the assigned site, so there is not need to define a StorageClass per zone or allowedTopologies there.
